File: open.kattis/triolingopush/triolingopush.py
from copy import deepcopy

def matMul(A, B):
	rowsA = len(A)
	colsA = len(A[0])
	rowsB = len(B)
	colsB = len(B[0])

	result = [[0 for row in range(colsB)] for col in range(rowsA)]
	for i in range(rowsA):
		for j in range(colsB):
			for k in range(colsA):
				result[i][j] += (A[i][k] * B[k][j])%(1000000007)
	return result

def exp(mult, a, n):
	if n == 1:
		return a 
	if n == 2:
		return mult(a, a)
	
	t = deepcopy(a)
	n-=1
	while n>1:
		if n%2 == 1:
			t = mult(a, t)
			n-=1
		a = mult(a, a)
		n>>=1
	return mult(a, t)

def main():
	n = int(input())
	mod = int(1e9)+7

	F = [[0], [1], [1]]
	T = [[0, 1, 0], [1, 1, 1], [0, 0, 1]]
	
	nT = exp(matMul, T, n)

	print(matMul(nT, F)[0][0]%mod)
main()

File: open.kattis/triolingopush/triolingopush.cpp
#include <bits/stdc++.h>
using namespace std;

typedef long long ll;
constexpr auto mod = ll(1e9 + 7);

struct Mod {
  ll x;
  Mod() : x(0) {}
  Mod(ll xx) : x(xx) {}
  Mod operator+(Mod b) { return Mod((x + b.x) % mod); }
  Mod operator-(Mod b) { return Mod((x - b.x + mod) % mod); }
  Mod operator*(Mod b) { return Mod((x * b.x) % mod); }

  Mod operator^(ll e) {
    if (!e)
      return Mod(1);
    Mod r = *this ^ (e / 2);
    r = r * r;
    return e & 1 ? *this * r : r;
  }
};

typedef vector<Mod> vll;
typedef vector<vll> vvll;

vvll matMul(vvll A, vvll B) {
  // std::cout << "MatMul 1" << std::endl;
  int rowsA = A.size();
  int colsA = A[0].size();
  // int rowsB = B.size();
  int colsB = B[0].size();

  vvll res(rowsA, vll(colsB));
  for (int i = 0; i < rowsA; i++) {
    // std::cout << "MatMul 2" << std::endl;
    for (int j = 0; j < colsB; j++) {
      // std::cout << "MatMul 3" << std::endl;
      Mod v = 0;
      for (int k = 0; k < colsA; k++) {
        // std::cout << "MatMul 4" << std::endl;
        v = v + A[i][k] * B[k][j];
      }
      res[i][j] = v;
    }
  }
  return res;
}

vvll matPow(vvll A, long long n) {
  if (n == 1) {
    return A;
  } else if (n == 2) {
    return matMul(A, A);
  } else {
    // std::cout << "MatPow" << std::endl;
    vvll t(A);
    n--;
    while (n > 1) {
      if (n & 1) {
        t = matMul(A, t);
      }
      A = matMul(A, A);
      n >>= 1;
    }
    return matMul(A, t);
  }
}

int main() {
  long long n;
  std::cin >> n;

  vvll F = {{0}, {1}, {1}};
  vvll T = {{0, 1, 0}, {1, 1, 1}, {0, 0, 1}};

  vvll nT = matPow(T, n);

  Mod Fn = matMul(nT, F)[0][0];
  std::cout << Fn.x << std::endl;
}